internal/helm/helm.go

```diff
@@ -0,0 +1,59 @@
+package helm
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os"
+	"os/exec"
+
+	"gopkg.in/yaml.v3"
+)
+
+type PullOptions struct {
+	ChartURL  string
+	Version   string
+	OutputDir string
+}
+
+func Pull(ctx context.Context, opts PullOptions) error {
+	chartURL, err := url.Parse(opts.ChartURL)
+	if err != nil {
+		return fmt.Errorf("invalid chart url: %w", err)
+	}
+
+	if chartURL.Scheme == "" {
+		chartURL.Scheme = "oci"
+	}
+
+	if opts.OutputDir == "" {
+		opts.OutputDir = "."
+	}
+
+	args := []string{"pull", chartURL.String(), "--untar", "--untardir", opts.OutputDir}
+	if opts.Version != "" {
+		args = append(args, "--version", opts.Version)
+	}
+
+	cmd := exec.CommandContext(ctx, "helm", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	cmd.Stdin = os.Stdin
+
+	return cmd.Run()
+}
+
+func Render(ctx context.Context, w io.Writer, chartPath string, values map[string]any) error {
+	var input bytes.Buffer
+	if err := yaml.NewEncoder(&input).Encode(values); err != nil {
+		return err
+	}
+
+	cmd := exec.CommandContext(ctx, "helm", "template", "joy-release-render", chartPath, "--values", "-")
```

Comment:
For my education, what's the benefit of carrying the context here?

Reply:
In this particular case, none - but only because we haven't added signal cancelation to our root context yet.
When we do, when we hit ctl+c instead of forcefully terminating the program, we can cancel our context and give the program a chance to:
- cleanup whatever its doing.
  - finish writing its file
  - close a network connection
- More importantly - return an error about what happened when it was canceled
  - failed to render chart: context canceled: interrupt
  - failed to pull chart: context canceled: interrupt

As a general rule of good practice, in Go, its important to not shy away from contexts.

Reply:
Yep, I should have read the docs:

```
package main

import (
	"context"
	"os/exec"
	"time"
)

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
	defer cancel()

	if err := exec.CommandContext(ctx, "sleep", "5").Run(); err != nil {
		// This will fail after 100 milliseconds. The 5 second sleep
		// will be interrupted.
	}
}

```